File: =EXAMTIME/1_TradeDesk/Word Search top to bottom1s/lc.java
import java.util.ArrayList;
import java.util.List;

public class lc {
    public static int solveLC(char[][] mat, String[] words) {
        int m = mat.length;
        int n = mat[0].length;
        List<Integer> res = new ArrayList<>();

        for (String w : words) {
            int cnt = 0;
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    for (int[] dir : new int[][]{{0, 1}, {1, 0}}) {
                        int di = dir[0];
                        int dj = dir[1];
                        int x = i, y = j;

                        if (w.charAt(0) != mat[x][y]) continue;

                        for (int k = 1; k < w.length(); k++) {
                            x += di;
                            y += dj;

                            if (x < 0 || y < 0 || x == m || y == n) break;

                            if (w.charAt(k) != mat[x][y]) {
                                break;
                            }

                            if (x == m - 1 || y == n - 1) {
                                di = -di;
                                dj = -dj;
                            }
                        }

                        if (x == m - 1 || y == n - 1) {
                            cnt++;
                        }
                    }
                }
            }

            res.add(cnt);
        }

        return res.stream().mapToInt(Integer::intValue).sum();
    }

    public static void main(String[] args) {
        char[][] matrix = {
                {'a', 'b', 'a', 'c'},
                {'x', 'a', 'c', 'd'},
                {'y', 'r', 'd', 's'}
        };

        String[] words = {"ac", "cat", "car", "bar", "acdc", "abacaba"};

        System.out.println(solveLC(matrix, words));  // Output: 7
    }
}

File: =EXAMTIME/1_TradeDesk/Word Search top to bottom1s/lc.py
def solveLC(mat, words):
    m,n = len(mat), len(mat[0])
    res = []
    for w in words:
        cnt = 0
        for i in range(len(mat)):
            for j in range(len(mat[0])):
                for di,dj in [(0,1),(1,0)]:
                    x,y = i,j
                    if w[0] != mat[x][y]: continue
                    for k in range(1,len(w)):
                        x += di
                        y += dj
                        if x < 0 or y < 0 or x == m or y == n: break
                        if w[k] != mat[x][y]:
                            break
                        if x == m-1 or y == n-1:
                            di,dj = -di,-dj
                    else:
                        cnt += 1

        res.append(cnt)
    return sum(res)
# print(solveLC([['a','a','a'],['a','a','a']],['aaaa'])) 
# print(solveLC([['a','a','a'],['a','a','a']],['aaa']))
print(solveLC([['a','b','a','c'],['x','a','c','d'],['y','r','d','s']],["ac", "cat", "car", "bar", "acdc", "abacaba"]))
# print(solveLC([['a','b','a','c'],['x','a','c','d'],['y','r','d','s']],["ac"]))